src/syntax.jl

```diff
@@ -0,0 +1,143 @@
+export analyze_syntax
+
+#####
+##### SyntaxTrees
+#####
+
+# Avoid piracy by defining AbstractTrees definitions on a wrapper
+struct SyntaxNodeWrapper
+    node::JuliaSyntax.SyntaxNode
+end
+
+function AbstractTrees.children(wrapper::SyntaxNodeWrapper)
+    # Don't recurse into these, in order to try to count top-level objects only
+    if kind(wrapper.node.raw) in [ K"struct", K"call", K"quote", K"=", K"for", K"function"]
+        return ()
+    else
+        map(SyntaxNodeWrapper, JuliaSyntax.children(wrapper.node))
+    end
+end
+
+#####
+##### Parsing files, traversing literal include statements
+#####
+
+function parse_syntax_one(file_path)
+    file = read(file_path, String)
+    parsed = JuliaSyntax.parse(JuliaSyntax.SyntaxNode, file)
+    return SyntaxNodeWrapper(parsed[1])
+end
+
+
+function find_literal_includes(tree::SyntaxNodeWrapper)
+    items = PostOrderDFS(tree)
+
+    # Filter to includes
+    itr = Iterators.filter(items) do wrapper
+        # A literal include has kind `K"call"`, whose first children's value is the symbol `:include`,
+        # and whose second child is the thing being included, which we require to be a string literal
+        k = kind(wrapper.node.raw)
+        k == K"call" || return false
+        kids = JuliaSyntax.children(wrapper.node)
+        isempty(kids) && return false
+        if kids[1].val == Symbol(:include) && kind(kids[2].raw) == K"string"
+            return true
+        else
+            return false
+        end
+        return true
+    end
+
+    # Grab the filenames
+    return map(itr) do wrapper
+        kids = JuliaSyntax.children(wrapper.node)
+        return only(kids[2].val).val
```

Comment:
Could probably fail for `include("file/$name.jl")` ?

Reply:
Yeah, that's why I called it `find_literal_includes` (since that call isn't using a string literal). I don't think PackageAnalyzer wants to go beyond static analysis, so at least for now I was thinking to just not worry about this case.

The other option we have is to just analyze all `.jl` files in the src directory, regardless if they know they are included or not. I think that would be a pretty reasonable choice as well and maybe I should switch to it.

Reply:
I think it's fine to stick to static analysis.  But my point is that the `filter` call above looks like it may think that `include("file/$name.jl")` is a literal. But it's not - it contains an interpolation.

Reply:
I mean, it's also fine imo to just analyze all `.jl` files in the src directory. Either way seems good :-)

The fact that `include` is dynamic is pretty problematic for static analysis tools in general :grimacing: